### ISSUE TITLE
DNS Leak in VPN mode

### ISSUE THREAD

**Issue by robin98** [LorenEteval/Furious#52]
## Describe the problem

While using VPN mode in windows, I noticed this program have serious DNS Leak issue.
originally while I'm in VPN mode, all my requests including DNS traffic must direct to the selected server. It done well in system proxy mode, but I detect DNS Leak in VPN mode.

## How to reproduce

- update my subscription
- set the routing as `Global`
- enabling VPN Mode via program
- after connection established, I open the CMD and check some blocked domain (by GFW) using ping command, it'll blocked like you never run VPN!

## Screenshot
![image](https://github.com/LorenEteval/Furious/assets/22413785/d79c4078-0e5d-4a29-9f8e-14599e32d02d)

**Comment by LorenEteval** [comment]
Hi. What you mean by blocked? The ping is successful in your screenshot.

**Comment by robin98** [comment]
> Hi. What you mean by blocked? The ping is successful in your screenshot.

No, check the IP address. it is NOT the twitter. 
it is a Private range IP address and every time we see this (in Iran) it means GFW somehow sniffed DNS traffic.
In this case, they can see that I sending request to a blocked website, and returns that IP address instead of actuall IP of twitter (DNS Transparent Proxy).

**Comment by LorenEteval** [comment]
> > Hi. What you mean by blocked? The ping is successful in your screenshot.
> 
> No, check the IP address. it is NOT the twitter. it is a Private range IP address and every time we see this (in Iran) it means GFW somehow sniffed DNS traffic. In this case, they can see that I sending request to a blocked website, and returns that IP address instead of actuall IP of twitter (DNS Transparent Proxy).

Please note:

The VPN mode is actually a re-name on TUN mode as it is written clearly in the wiki, which is based on [tun2socks](https://github.com/xjasonlyu/tun2socks) which can only proxy L4 traffic, while ping is in L3.

Previously **the** tun2socks does not handle ping traffic. Then the ping traffic is handled by simple echo back in the **localhost (127.0.0.1)**. Please check if your environment is setup correctly.

**Comment by LorenEteval** [comment]
> > > Hi. What you mean by blocked? The ping is successful in your screenshot.
> > 
> > 
> > No, check the IP address. it is NOT the twitter. it is a Private range IP address and every time we see this (in Iran) it means GFW somehow sniffed DNS traffic. In this case, they can see that I sending request to a blocked website, and returns that IP address instead of actuall IP of twitter (DNS Transparent Proxy).
> 
> Please note:
> 
> The VPN mode is actually a re-name on TUN mode as it is written clearly in the wiki, which is based on [tun2socks](https://github.com/xjasonlyu/tun2socks) which can only proxy L4 traffic, while ping is in L3.
> 
> Previously **the** tun2socks does not handle ping traffic. ~Then the ping traffic is handled by simple echo back in the **localhost (127.0.0.1)**. Please check if your environment is setup correctly~.
> 
> Updated: After a few different censored website tests I can confirm that tun2socks proxy DNS traffic correctly. Please check if the VPN mode is actually enabled. You can also do more tests suggested in the wiki page.

**Comment by LorenEteval** [comment]
> > Hi. What you mean by blocked? The ping is successful in your screenshot.
> 
> No, check the IP address. it is NOT the twitter. it is a Private range IP address and every time we see this (in Iran) it means GFW somehow sniffed DNS traffic. In this case, they can see that I sending request to a blocked website, and returns that IP address instead of actuall IP of twitter (DNS Transparent Proxy).

This issue seems to be related to the upstream, i.e. tun2socks. I'm not sure if there is anything to do on the application side.

**Comment by LorenEteval** [comment]
Can you try again with `curl` suggested in the [wiki page verify](https://github.com/LorenEteval/Furious/wiki/TUN-Mode#verify)? If there's DNS leak then the `curl` test will never pass in Iran.

**Comment by robin98** [comment]
> Can you try again with `curl` suggested in the [wiki page verify](https://github.com/LorenEteval/Furious/wiki/TUN-Mode#verify)? If there's DNS leak then the `curl` test will never pass in Iran.

It seems that the `curl` can pass the test, but still when I make a `ping` test, it'll return me the blocked IP.
This case became even more starnge when I noticed there is no problem with `nslookup` command here!

This problem does not happen with my other client, [hiddify-Next](https://github.com/hiddify/hiddify-next) for desktop.

I can't confirm, but I think this kind of possible DNS Leak in VPN mode, causing future problem in the software such Android studio.

**Comment by LorenEteval** [comment]
> > Can you try again with `curl` suggested in the [wiki page verify](https://github.com/LorenEteval/Furious/wiki/TUN-Mode#verify)? If there's DNS leak then the `curl` test will never pass in Iran.
> 
> It seems that the `curl` can pass the test, but still when I make a `ping` test, it'll return me the blocked IP. This case became even more starnge when I noticed there is no problem with `nslookup` command here!
> 
> This problem does not happen with my other client, [hiddify-Next](https://github.com/hiddify/hiddify-next) for desktop.
> 
> I can't say this for sure, but I think this kind of possible DNS Leak in VPN mode, causing future problem in the software such Android studio.

Have you tried other domains besides twitter? Did they also show incorrect ip address during ping test?

**Comment by robin98** [comment]
> > > Can you try again with `curl` suggested in the [wiki page verify](https://github.com/LorenEteval/Furious/wiki/TUN-Mode#verify)? If there's DNS leak then the `curl` test will never pass in Iran.
> > 
> > 
> > It seems that the `curl` can pass the test, but still when I make a `ping` test, it'll return me the blocked IP. This case became even more starnge when I noticed there is no problem with `nslookup` command here!
> > This problem does not happen with my other client, [hiddify-Next](https://github.com/hiddify/hiddify-next) for desktop.
> > I can't say this for sure, but I think this kind of possible DNS Leak in VPN mode, causing future problem in the software such Android studio.
> 
> Have you tried other domains besides twitter? Did they also show incorrect ip address during ping test?

Yes, I do. I also tried youtube, facebook and etc... all of them show the signs of DNS Leakage.

**Comment by LorenEteval** [comment]
> > > > Can you try again with `curl` suggested in the [wiki page verify](https://github.com/LorenEteval/Furious/wiki/TUN-Mode#verify)? If there's DNS leak then the `curl` test will never pass in Iran.
> > > 
> > > 
> > > It seems that the `curl` can pass the test, but still when I make a `ping` test, it'll return me the blocked IP. This case became even more starnge when I noticed there is no problem with `nslookup` command here!
> > > This problem does not happen with my other client, [hiddify-Next](https://github.com/hiddify/hiddify-next) for desktop.
> > > I can't say this for sure, but I think this kind of possible DNS Leak in VPN mode, causing future problem in the software such Android studio.
> > 
> > 
> > Have you tried other domains besides twitter? Did they also show incorrect ip address during ping test?
> 
> Yes, I do. I also tried youtube, facebook and etc... all of them show the signs of DNS Leakage.

OK. I can confirm that Furious has DNS leak by analyzing traffic in the Wireshark, and I know where the problem is and how to fix it. I guess that's a first-priority fix.

I'll keep this issue open until it is confirmed fixed. Would you mind having another test in the next version to see if the issue still exists? Thank you.

**Comment by robin98** [comment]
> OK. I can confirm that Furious has DNS leak by analyzing traffic in the Wireshark, and I know where the problem is and how to fix it. I guess that's a first-priority fix.
> 
> I'll keep this issue open until it is confirmed fixed. Would you mind having another test in the next version to see if the issue still exists? Thank you. Since I'm very busy recently, I'll just add a prerelease based on old code base focusing on just this issue(no new features, etc.). Is it OK to you?
> 
> Updated: observed on Windows. Not sure if macOS version has this issue.

Thank you. this is good news.
I will wait for you to publish it.

**Comment by robin98** [comment]
I appreciate for this. It seems that the problem has been fixed.
